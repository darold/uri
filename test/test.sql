DROP DATABASE test_uri;
CREATE DATABASE test_uri;

\c test_uri
CREATE EXTENSION uri;

CREATE TABLE t1 (id serial, url uri);
INSERT INTO t1 (url) VALUES ('http://pgcluu.darold.net/index.html');
INSERT INTO t1 (url) VALUES ('file:///opt/git/uri/README');
INSERT INTO t1 (url) VALUES ('http://mydomain.com/pub/../public/../index 1.html');
INSERT INTO t1 (url) VALUES ('http://username:passwd@mydomain.com/pub/index.php');
INSERT INTO t1 (url) VALUES ('http://192.168.1.1/');
INSERT INTO t1 (url) VALUES ('http://pgcluu.darold.net/example/pgbouncer-dolibarr.html#pgbouncer-duration');
INSERT INTO t1 (url) VALUES ('ftp://ftp.is.co.za/rfc/rfc1808.txt');
INSERT INTO t1 (url) VALUES ('http://www.ietf.org/rfc/rfc2396.txt');
INSERT INTO t1 (url) VALUES ('ldap://[2001:db8::7]/c=GB?objectClass?one');
INSERT INTO t1 (url) VALUES ('mailto:John.Doe@test.com');
INSERT INTO t1 (url) VALUES ('news:comp.infosystems.www.servers.unix');
INSERT INTO t1 (url) VALUES ('tel:+1-816-555-1212');
INSERT INTO t1 (url) VALUES ('telnet://192.0.2.16:80/');
INSERT INTO t1 (url) VALUES ('urn:oasis:names:specification:docbook:dtd:xml:4.1.2');
INSERT INTO t1 (url) VALUES ('switch?toggle=1');
INSERT INTO t1 (url) VALUES ('file:///etc/postgresql/9.6/main/postgresql.conf');
INSERT INTO t1 (url) VALUES ('file:///etc/postgresql/9.3/main/../../9.6/main/postgresql.conf');
INSERT INTO t1 (url) VALUES ('/etc/postgresql/9.6/main/postgresql.conf');
-- Normalize URIs according to RFC 3986
INSERT INTO t1 (url) VALUES ('HTTPS://www.Example.com:443/../test/../foo/index.html');
INSERT INTO t1 (url) VALUES ('https://WWW.EXAMPLE.COM/./foo/index.html');
INSERT INTO t1 (url) VALUES ('https://www.example.com/%66%6f%6f/index.html');
INSERT INTO t1 (url) VALUES ('https://www.example.com/foo/index.html');

-- Test index creation
CREATE INDEX t1_btree_index_uri ON t1 USING btree(url);
ANALYZE t1;

-- Get all records
SELECT * FROM t1;

-- Use btree index
SET enable_seqscan = off;
EXPLAIN (COSTS off) SELECT * FROM t1 WHERE url = 'file:///etc/postgresql/9.3/main/../../9.6/main/postgresql.conf';
SET enable_seqscan = on;

-- Test all functions of the uri extension
\x on
SELECT url as uri,
        uri_get_scheme(url) as scheme,
        uri_get_auth(url) as authen,
        uri_get_host(url) as host,
        uri_get_port(url) as port,
        uri_get_portnum(url) as port,
        uri_get_path(url) as path,
        uri_get_query(url) as query,
        uri_get_fragment(url) as fragment
  FROM t1;

-- Test all extended functions
SELECT
	uri_get_str('http://192.168.1.1/index.html') as string,
	uri_is_absolute('http://192.168.1.1/index.html') as abs_true,
	uri_is_absolute('/index.html') as abs_false,
	uri_is_absolute_path('/index.html') as abs_path_true,
	uri_is_absolute_path('index.html') as abs_path_false,
	uri_localpath_exists('/etc/fstab') as local_exists_true,
	uri_localpath_exists('/etc/fstab.no') as local_exists_false,
	uri_remotepath_exists('http://ora2pg.darold.net/index.html') as remotepath_true,
	uri_remotepath_exists('http://ora2pg.darold.net/index2.html') as remotepath_false,
	uri_path_exists('http://ora2pg.darold.net/index.html') as rpath_exists_true,
	uri_path_exists('http://ora2pg.darold.net/index2.html') as rpath_exists_false,
	uri_path_exists('/etc/fstab') as lpath_exists_true,
	uri_path_exists('/etc/motd') as lpath_exists_false, -- Symlink
	uri_localpath_content_type('/etc/fstab') as local_content_type,
	uri_remotepath_content_type('http://ora2pg.darold.net/index.html') as remote_content_type,
	uri_path_content_type('/etc/fstab') as lpath_content_type,
	uri_path_content_type('https://avatars2.githubusercontent.com/u/538862?s=40&v=4') as rpath_content_type,
	uri_localpath_size('/etc/fstab') as local_size,
	uri_remotepath_size('http://ora2pg.darold.net/index.html') as remote_size,
	uri_path_size('/etc/fstab') as lpath_size,
	uri_path_size('http://www.darold.net/confs/pgday_2017_partitionnement.pdf') as rpath_size;
\x off

-- Test sorting
SELECT DISTINCT url FROM t1 ORDER BY url;
-- Test sort using index only scan
SET enable_seqscan = off;
EXPLAIN (COSTS off) SELECT DISTINCT url FROM t1 ORDER BY url;
SET enable_seqscan = on;

-- Test hash join
CREATE TABLE t2 (lbl text, url uri);
INSERT INTO t2 VALUES ('mybox', 'http://192.168.1.1/');
ANALYZE t2;
SET enable_nestloop = off;
SET enable_mergejoin = off;
SELECT * FROM t1 JOIN t2 ON t1.url = t2.url AND t1.id = 5;
EXPLAIN (COSTS off) SELECT * FROM t1 JOIN t2 ON t1.url = t2.url AND t1.id = 5;
SET enable_nestloop = on;
SET enable_mergejoin = on;

-- Test hash index
DROP INDEX t1_btree_index_uri;
CREATE INDEX t1_hash_index_uri  ON t1 USING hash(url);
ANALYZE t1;
SET enable_seqscan = off;
SET enable_hashjoin = off;
SET enable_mergejoin = off;
SELECT * FROM t1 JOIN t2 ON t1.url = t2.url AND t1.id = 5;
EXPLAIN (COSTS off) SELECT * FROM t1 JOIN t2 ON t1.url = t2.url AND t1.id = 5;
SET enable_seqscan = on;
SET enable_hashjoin = on;
SET enable_mergejoin = on;

-- Test contain operators
SELECT * FROM t1 WHERE url @> '192.168.1.1';
SELECT * FROM t1 WHERE '192.168.1.1' <@ url;
SELECT * FROM t1 WHERE 'http://192.168.1.1/index.html' @> url;
SELECT * FROM t1 WHERE url <@ 'http://192.168.1.1/index.html';

