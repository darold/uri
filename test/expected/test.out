DROP DATABASE
CREATE DATABASE
You are now connected to database "test_uri" as user "gilles".
CREATE EXTENSION
CREATE TABLE
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
INSERT 0 1
CREATE INDEX
ANALYZE
 id |                                     url                                     
----+-----------------------------------------------------------------------------
  1 | http://pgcluu.darold.net/index.html
  2 | file:///opt/git/uri/README
  3 | http://mydomain.com/index%201.html
  4 | http://username:passwd@mydomain.com/pub/index.php
  5 | http://192.168.1.1/
  6 | http://pgcluu.darold.net/example/pgbouncer-dolibarr.html#pgbouncer-duration
  7 | ftp://ftp.is.co.za/rfc/rfc1808.txt
  8 | http://www.ietf.org/rfc/rfc2396.txt
  9 | ldap://[2001:0db8:0000:0000:0000:0000:0000:0007]/c=GB?objectClass?one
 10 | mailto:John.Doe@test.com
 11 | news:comp.infosystems.www.servers.unix
 12 | tel:+1-816-555-1212
 13 | telnet://192.0.2.16:80/
 14 | urn:oasis:names:specification:docbook:dtd:xml:4.1.2
 15 | switch?toggle=1
 16 | file:///etc/postgresql/9.6/main/postgresql.conf
 17 | file:///etc/postgresql/9.6/main/postgresql.conf
 18 | /etc/postgresql/9.6/main/postgresql.conf
 19 | https://www.example.com:443/foo/index.html
 20 | https://www.example.com/foo/index.html
 21 | https://www.example.com/foo/index.html
 22 | https://www.example.com/foo/index.html
(22 rows)

SET
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Index Scan using t1_btree_index_uri on t1
   Index Cond: (url = 'file:///etc/postgresql/9.6/main/postgresql.conf'::uri)
(2 rows)

SET
Expanded display is on.
-[ RECORD 1 ]-------------------------------------------------------------------------
uri      | http://pgcluu.darold.net/index.html
scheme   | http
authen   | 
host     | pgcluu.darold.net
port     | 
port     | 0
path     | /index.html
query    | 
fragment | 
-[ RECORD 2 ]-------------------------------------------------------------------------
uri      | file:///opt/git/uri/README
scheme   | file
authen   | 
host     | 
port     | 
port     | 0
path     | /opt/git/uri/README
query    | 
fragment | 
-[ RECORD 3 ]-------------------------------------------------------------------------
uri      | http://mydomain.com/index%201.html
scheme   | http
authen   | 
host     | mydomain.com
port     | 
port     | 0
path     | /index%201.html
query    | 
fragment | 
-[ RECORD 4 ]-------------------------------------------------------------------------
uri      | http://username:passwd@mydomain.com/pub/index.php
scheme   | http
authen   | username:passwd
host     | mydomain.com
port     | 
port     | 0
path     | /pub/index.php
query    | 
fragment | 
-[ RECORD 5 ]-------------------------------------------------------------------------
uri      | http://192.168.1.1/
scheme   | http
authen   | 
host     | 192.168.1.1
port     | 
port     | 0
path     | /
query    | 
fragment | 
-[ RECORD 6 ]-------------------------------------------------------------------------
uri      | http://pgcluu.darold.net/example/pgbouncer-dolibarr.html#pgbouncer-duration
scheme   | http
authen   | 
host     | pgcluu.darold.net
port     | 
port     | 0
path     | /example/pgbouncer-dolibarr.html
query    | 
fragment | pgbouncer-duration
-[ RECORD 7 ]-------------------------------------------------------------------------
uri      | ftp://ftp.is.co.za/rfc/rfc1808.txt
scheme   | ftp
authen   | 
host     | ftp.is.co.za
port     | 
port     | 0
path     | /rfc/rfc1808.txt
query    | 
fragment | 
-[ RECORD 8 ]-------------------------------------------------------------------------
uri      | http://www.ietf.org/rfc/rfc2396.txt
scheme   | http
authen   | 
host     | www.ietf.org
port     | 
port     | 0
path     | /rfc/rfc2396.txt
query    | 
fragment | 
-[ RECORD 9 ]-------------------------------------------------------------------------
uri      | ldap://[2001:0db8:0000:0000:0000:0000:0000:0007]/c=GB?objectClass?one
scheme   | ldap
authen   | 
host     | 2001:0db8:0000:0000:0000:0000:0000:0007
port     | 
port     | 0
path     | /c=GB
query    | objectClass?one
fragment | 
-[ RECORD 10 ]------------------------------------------------------------------------
uri      | mailto:John.Doe@test.com
scheme   | mailto
authen   | 
host     | 
port     | 
port     | 0
path     | /John.Doe@test.com
query    | 
fragment | 
-[ RECORD 11 ]------------------------------------------------------------------------
uri      | news:comp.infosystems.www.servers.unix
scheme   | news
authen   | 
host     | 
port     | 
port     | 0
path     | /comp.infosystems.www.servers.unix
query    | 
fragment | 
-[ RECORD 12 ]------------------------------------------------------------------------
uri      | tel:+1-816-555-1212
scheme   | tel
authen   | 
host     | 
port     | 
port     | 0
path     | /+1-816-555-1212
query    | 
fragment | 
-[ RECORD 13 ]------------------------------------------------------------------------
uri      | telnet://192.0.2.16:80/
scheme   | telnet
authen   | 
host     | 192.0.2.16
port     | 80
port     | 80
path     | /
query    | 
fragment | 
-[ RECORD 14 ]------------------------------------------------------------------------
uri      | urn:oasis:names:specification:docbook:dtd:xml:4.1.2
scheme   | urn
authen   | 
host     | 
port     | 
port     | 0
path     | /oasis:names:specification:docbook:dtd:xml:4.1.2
query    | 
fragment | 
-[ RECORD 15 ]------------------------------------------------------------------------
uri      | switch?toggle=1
scheme   | 
authen   | 
host     | 
port     | 
port     | 0
path     | switch
query    | toggle=1
fragment | 
-[ RECORD 16 ]------------------------------------------------------------------------
uri      | file:///etc/postgresql/9.6/main/postgresql.conf
scheme   | file
authen   | 
host     | 
port     | 
port     | 0
path     | /etc/postgresql/9.6/main/postgresql.conf
query    | 
fragment | 
-[ RECORD 17 ]------------------------------------------------------------------------
uri      | file:///etc/postgresql/9.6/main/postgresql.conf
scheme   | file
authen   | 
host     | 
port     | 
port     | 0
path     | /etc/postgresql/9.6/main/postgresql.conf
query    | 
fragment | 
-[ RECORD 18 ]------------------------------------------------------------------------
uri      | /etc/postgresql/9.6/main/postgresql.conf
scheme   | 
authen   | 
host     | 
port     | 
port     | 0
path     | /etc/postgresql/9.6/main/postgresql.conf
query    | 
fragment | 
-[ RECORD 19 ]------------------------------------------------------------------------
uri      | https://www.example.com:443/foo/index.html
scheme   | https
authen   | 
host     | www.example.com
port     | 443
port     | 443
path     | /foo/index.html
query    | 
fragment | 
-[ RECORD 20 ]------------------------------------------------------------------------
uri      | https://www.example.com/foo/index.html
scheme   | https
authen   | 
host     | www.example.com
port     | 
port     | 0
path     | /foo/index.html
query    | 
fragment | 
-[ RECORD 21 ]------------------------------------------------------------------------
uri      | https://www.example.com/foo/index.html
scheme   | https
authen   | 
host     | www.example.com
port     | 
port     | 0
path     | /foo/index.html
query    | 
fragment | 
-[ RECORD 22 ]------------------------------------------------------------------------
uri      | https://www.example.com/foo/index.html
scheme   | https
authen   | 
host     | www.example.com
port     | 
port     | 0
path     | /foo/index.html
query    | 
fragment | 

-[ RECORD 1 ]-------+------------------------------
string              | http://192.168.1.1/index.html
abs_true            | t
abs_false           | f
abs_path_true       | t
abs_path_false      | f
local_exists_true   | t
local_exists_false  | f
remotepath_true     | t
remotepath_false    | f
rpath_exists_true   | t
rpath_exists_false  | f
lpath_exists_true   | t
lpath_exists_false  | f
local_content_type  | text/plain; charset=us-ascii
remote_content_type | text/html
lpath_content_type  | text/plain; charset=us-ascii
rpath_content_type  | image/png
local_size          | 674
remote_size         | 5098
lpath_size          | 674
rpath_size          | 101596

Expanded display is off.
                                     url                                     
-----------------------------------------------------------------------------
 /etc/postgresql/9.6/main/postgresql.conf
 file:///etc/postgresql/9.6/main/postgresql.conf
 file:///opt/git/uri/README
 ftp://ftp.is.co.za/rfc/rfc1808.txt
 http://192.168.1.1/
 http://mydomain.com/index%201.html
 http://pgcluu.darold.net/example/pgbouncer-dolibarr.html#pgbouncer-duration
 http://pgcluu.darold.net/index.html
 http://username:passwd@mydomain.com/pub/index.php
 http://www.ietf.org/rfc/rfc2396.txt
 https://www.example.com/foo/index.html
 https://www.example.com:443/foo/index.html
 ldap://[2001:0db8:0000:0000:0000:0000:0000:0007]/c=GB?objectClass?one
 mailto:John.Doe@test.com
 news:comp.infosystems.www.servers.unix
 switch?toggle=1
 tel:+1-816-555-1212
 telnet://192.0.2.16:80/
 urn:oasis:names:specification:docbook:dtd:xml:4.1.2
(19 rows)

SET
                      QUERY PLAN                      
------------------------------------------------------
 Unique
   ->  Index Only Scan using t1_btree_index_uri on t1
(2 rows)

SET
CREATE TABLE
INSERT 0 1
ANALYZE
SET
SET
 id |         url         |  lbl  |         url         
----+---------------------+-------+---------------------
  5 | http://192.168.1.1/ | mybox | http://192.168.1.1/
(1 row)

           QUERY PLAN           
--------------------------------
 Hash Join
   Hash Cond: (t1.url = t2.url)
   ->  Seq Scan on t1
         Filter: (id = 5)
   ->  Hash
         ->  Seq Scan on t2
(6 rows)

SET
SET
DROP INDEX
CREATE INDEX
ANALYZE
SET
SET
SET
 id |         url         |  lbl  |         url         
----+---------------------+-------+---------------------
  5 | http://192.168.1.1/ | mybox | http://192.168.1.1/
(1 row)

                   QUERY PLAN                   
------------------------------------------------
 Nested Loop
   ->  Seq Scan on t2
   ->  Index Scan using t1_hash_index_uri on t1
         Index Cond: (url = t2.url)
         Filter: (id = 5)
(5 rows)

SET
SET
SET
 id |         url         
----+---------------------
  5 | http://192.168.1.1/
(1 row)

 id |         url         
----+---------------------
  5 | http://192.168.1.1/
(1 row)

 id |         url         
----+---------------------
  5 | http://192.168.1.1/
(1 row)

 id |         url         
----+---------------------
  5 | http://192.168.1.1/
(1 row)

                                             uri_escape                                              
-----------------------------------------------------------------------------------------------------
 %24%20%26%20%3C%20%3E%20%3F%20%3B%20%23%20%3A%20%3D%20%2C%20%22%20%27%20~%20%2B%20%25%20%5Cr%20%5Cn
(1 row)

            uri_unescape             
-------------------------------------
 $ & < > ? ; # : = , " ' ~ + % \r \n
(1 row)

    uri_escape     
-------------------
 Test%0Anew%20line
(1 row)

 uri_unescape 
--------------
 Test        +
 new line
(1 row)

